File: routers/address_labels.py
"""Admin-curated address → name labels.

Global-plus-override model (invariant 12): a label row is either *global*
(``chain IS NULL`` — applies on every chain, the right semantics for EOA /
Safe-signer accounts) or *chain-qualified* (a concrete chain name that overrides
the global label on that chain, so contract labels are safe cross-chain). The
API preserves the historical address-keyed shape for global rows so existing
consumers are unchanged, and exposes chain-qualified rows under a separate map.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select

from db.models import AddressLabel
from schemas.api_requests import AddressLabelUpsert
from schemas.api_responses import (
    AddressLabelDeleteResponse,
    AddressLabelsResponse,
    AddressLabelUpsertResponse,
    AddressLabelView,
)
from utils.chains import UnknownChainError, chain_by_name

from . import deps

router = APIRouter()


def _resolve_chain_or_400(chain: str | None) -> str | None:
    """Normalize an optional ``?chain=`` param to a canonical chain name.

    ``None`` / empty stays ``None`` (operate on the global row — unchanged
    behavior). A concrete value is normalized through the registry so aliases
    (``"mainnet"`` → ``"ethereum"``) collapse to one row; an unknown chain is a
    client error (400) with the offending value, never a silent fallthrough.
    """
    if chain is None or not chain.strip():
        return None
    try:
        return chain_by_name(chain).name
    except UnknownChainError:
        raise HTTPException(status_code=400, detail=f"Unknown chain: {chain!r}") from None


def _row_view(row: AddressLabel) -> AddressLabelView:
    return {
        "name": row.name,
        "note": row.note,
        "updated_at": row.updated_at.isoformat() if row.updated_at else None,
    }


@router.get("/api/address_labels", response_model=None)
def list_address_labels() -> AddressLabelsResponse:
    """Return stored address → name mappings.

    ``labels`` keeps the original ``{address: {...}}`` shape and carries only
    GLOBAL rows, so every existing consumer is byte-compatible. Chain-qualified
    overrides live under ``chain_labels`` as ``{chain: {address: {...}}}``.

    Public read endpoint so any page (principal detail, surface node, etc.) can
    decorate raw hex addresses with the admin-assigned name. The admin key is
    only required to mutate labels (PUT/DELETE below).
    """
    with deps.SessionLocal() as session:
        rows = session.execute(select(AddressLabel)).scalars().all()
        labels: dict[str, AddressLabelView] = {}
        chain_labels: dict[str, dict[str, AddressLabelView]] = {}
        for row in rows:
            if row.chain is None:
                labels[row.address] = _row_view(row)
            else:
                chain_labels.setdefault(row.chain, {})[row.address] = _row_view(row)
        return {"labels": labels, "chain_labels": chain_labels}


@router.put("/api/address_labels/{address}", dependencies=[Depends(deps.require_admin_key)], response_model=None)
def upsert_address_label(
    address: str,
    payload: AddressLabelUpsert,
    chain: str | None = Query(default=None),
) -> AddressLabelUpsertResponse:
    """Create or update the human-readable name for an address.

    Idempotent — repeated calls with the same body leave the row unchanged
    (aside from ``updated_at``). Without ``?chain=`` this writes the GLOBAL row
    (``chain IS NULL``), exactly as before — the frontend uses that for Safe
    signers and EOA principals. With ``?chain=`` it writes the chain-qualified
    override for a contract on that specific network.
    """
    a = deps._normalize_address_or_400(address)
    c = _resolve_chain_or_400(chain)
    with deps.SessionLocal() as session:
        # Cannot key by PK (surrogate id now); match on the (address, chain)
        # override slot, with chain IS NULL selecting the single global row.
        stmt = select(AddressLabel).where(AddressLabel.address == a)
        stmt = stmt.where(AddressLabel.chain == c) if c is not None else stmt.where(AddressLabel.chain.is_(None))
        row = session.execute(stmt).scalar_one_or_none()
        if row is None:
            row = AddressLabel(address=a, chain=c, name=payload.name.strip(), note=payload.note)
            session.add(row)
        else:
            row.name = payload.name.strip()
            row.note = payload.note
        session.commit()
        deps.log_admin_mutation("address_label_upsert", id=a if c is None else f"{c}:{a}")
        return {
            "address": a,
            "chain": c,
            "name": row.name,
            "note": row.note,
            "updated_at": row.updated_at.isoformat() if row.updated_at else None,
        }


@router.delete("/api/address_labels/{address}", dependencies=[Depends(deps.require_admin_key)], response_model=None)
def delete_address_label(
    address: str,
    chain: str | None = Query(default=None),
) -> AddressLabelDeleteResponse:
    a = deps._normalize_address_or_400(address)
    c = _resolve_chain_or_400(chain)
    with deps.SessionLocal() as session:
        stmt = select(AddressLabel).where(AddressLabel.address == a)
        stmt = stmt.where(AddressLabel.chain == c) if c is not None else stmt.where(AddressLabel.chain.is_(None))
        row = session.execute(stmt).scalar_one_or_none()
        if row is None:
            raise HTTPException(status_code=404, detail="Label not found")
        session.delete(row)
        session.commit()
        deps.log_admin_mutation("address_label_delete", id=a if c is None else f"{c}:{a}")
        return {"address": a, "chain": c, "deleted": True}

File: site/src/api/addressLabels.js
// Thin API wrappers for /api/address_labels/*. See api.py for the server
// shapes. Admin mutations (put/delete) use the shared api() helper so the
// X-PSAT-Admin-Key header is auto-injected; a missing/invalid key triggers
// the built-in 401 prompt-retry flow.
//
// Global-plus-override model (invariant 12): a label is either GLOBAL (applies
// on every chain — the right thing for EOA/Safe-signer accounts) or
// CHAIN-QUALIFIED (overrides the global label for one network — the right thing
// for contracts, which are a different deployment at the same address per
// chain). Omit `chain` to operate on the global row; pass it for the override.

import { api } from "./client.js";

export function listAddressLabels() {
  return api("/api/address_labels");
}

export function upsertAddressLabel(address, name, note = null, chain = null) {
  const body = note == null ? { name } : { name, note };
  const q = chain ? `?chain=${encodeURIComponent(chain)}` : "";
  return api(`/api/address_labels/${encodeURIComponent(address)}${q}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

export function deleteAddressLabel(address, chain = null) {
  const q = chain ? `?chain=${encodeURIComponent(chain)}` : "";
  return api(`/api/address_labels/${encodeURIComponent(address)}${q}`, {
    method: "DELETE",
  });
}

// Build lookup maps from the /api/address_labels response. Returns
// `{ global: Map<addr,name>, byChain: Map<chain, Map<addr,name>> }`. Addresses
// are lowercased so lookups are case-insensitive. Tolerant of the legacy shape
// (a response with only `labels` and no `chain_labels`).
export function buildLabelMaps(resp) {
  const global = new Map();
  for (const [addr, row] of Object.entries(resp?.labels || {})) {
    global.set(String(addr).toLowerCase(), row?.name);
  }
  const byChain = new Map();
  for (const [chain, rows] of Object.entries(resp?.chain_labels || {})) {
    const m = new Map();
    for (const [addr, row] of Object.entries(rows || {})) {
      m.set(String(addr).toLowerCase(), row?.name);
    }
    byChain.set(chain, m);
  }
  return { global, byChain };
}

// Chain-specific-wins-else-global lookup. A chain-qualified label for
// (address, chain) overrides the global one; with no chain, or no override for
// that chain, the global label is returned. Missing -> null.
export function lookupLabel(maps, address, chain = null) {
  const addr = String(address || "").toLowerCase();
  if (chain) {
    const m = maps?.byChain?.get?.(chain);
    if (m && m.has(addr)) return m.get(addr) ?? null;
  }
  return maps?.global?.get?.(addr) ?? null;
}

// Resolve the display name from either shape AddressLabelInline may be handed:
// a plain Map<addr,name> (legacy global-only callers) or the maps struct from
// buildLabelMaps (chain-aware callers). Keeps the component's `labels` prop
// backward compatible.
export function resolveLabelName(labels, address, chain = null) {
  const addr = String(address || "").toLowerCase();
  if (labels && typeof labels.get === "function" && !("global" in labels)) {
    return labels.get(addr) ?? null;
  }
  return lookupLabel(labels, addr, chain);
}
